### PR TITLE
Implement RFC 873: Type Macros

### PR DESCRIPTION
**Pull request by jroesch**
This pull request implements the functionality for [RFC 873](https://github.com/rust-lang/rfcs/blob/master/text/0873-type-macros.md). This is currently just an update of @freebroccolo's branch from January, the corresponding commits are linked in each commit message.

@nikomatsakis and I had talked about updating the macro language to support a lifetime fragment specifier, and it is possible to do that work on this branch as well. If so we can (collectively) talk about it next week during the pre-RustCamp work week.

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Can you ad a feature gate (standard procedure)?

**Comment by jroesch** [comment]
@nikomatsakis yeah, I wasn't sure about how we were doing RFCs. Feature gate for one release cycle?

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -163,8 +163,12 @@ const KNOWN_FEATURES: &'static [(&'static str, &'static str, Status)] = &[
...
     // Allows the definition recursive static items.
     ("static_recursion", "1.3.0", Active),
-// Allows default type parameters to influence type inference.
-    ("default_type_parameter_fallback", "1.3.0", Active)
+
+    // Allows default type parameters to influence type inference.
+    ("default_type_parameter_fallback", "1.3.0", Active),
+
+    // Allows macros to appear in the type position.
+    ("type_macros", "1.3.0", Active)
```
Comment (huonw) [inline src/libsyntax/feature_gate.rs]: r-

File: src/libsyntax/ext/expand.rs [hunk 2]
```diff
@@ -1555,6 +1555,45 @@ fn expand_and_rename_method(sig: ast::MethodSig, body: P<ast::Block>,
...
+                                                         mark_ty,
+                                                         fld) {
+                    Some(ty) => ty,
+                    None => {
+                        return DummyResult::raw_ty(t.span);
+                    }
+                };
+
+                // Keep going, outside-in.
+                //
```
Comment (huonw) [inline src/libsyntax/ext/expand.rs]: Empty trailing comment?

**Comment by huonw** [comment]
Could you add a compile-fail test for the feature gate? And maybe some more basic run-pass tests for the macros themselves?

File: src/libsyntax/ext/expand.rs [hunk 3]
```diff
@@ -1552,6 +1552,44 @@ fn expand_and_rename_method(sig: ast::MethodSig, body: P<ast::Block>,
...
+                    id: ast::DUMMY_NODE_ID,
+                    node: t.node,
+                    span: t.span,
+                    })
+            } else {
+                feature_gate::emit_feature_err(
+                    &fld.cx.parse_sess.span_diagnostic,
+                    "type_macros",
+                    t.span,
+                    "type macros are experimental (see tracking issue: 27336)");
```
Comment (huonw) [inline src/libsyntax/ext/expand.rs]: I think `(see issue #27336)` is a slightly better phrasing of this.

File: src/test/compile-fail/type-macros-fail.rs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+macro_rules! Id {
+    { $T:tt } => $T
+}
+
+struct Foo<T> {
+    x: Id!(T)
```
Comment (huonw) [inline src/test/compile-fail/type-macros-fail.rs]: Shouldn't this have an `// ~ERROR ...`?

File: src/libsyntax/feature_gate.rs [hunk 5]
```diff
@@ -169,6 +169,8 @@ const KNOWN_FEATURES: &'static [(&'static str, &'static str, Status)] = &[
 
     // Allows associated type defaults
     ("associated_type_defaults", "1.2.0", Active),
+// Allows macros to appear in the type position.
```
Comment (huonw) [inline src/libsyntax/feature_gate.rs]: Indentation

File: src/libsyntax/feature_gate.rs [hunk 6]
```diff
@@ -228,8 +230,7 @@ pub const KNOWN_ATTRIBUTES: &'static [(&'static str, AttributeType)] = &[
                      "no_std is experimental")),
     ("lang", Gated("lang_items",
                      "language items are subject to change")),
-    ("linkage", Gated("linkage",
-                      "the `linkage` attribute is experimental \
+    ("linkage", Gated("linkage", "the `linkage` attribute is experimental \
```
Comment (huonw) [inline src/libsyntax/feature_gate.rs]: Is this change really necessary? :P

**Comment by huonw** [comment]
r=me with an ERROR in the compile-fail test (and some nits).

**Comment by jroesch** [comment]
I addressed all your comments and it should be good to go. @bors r=huonw
